src/main/java/com/linecorp/armeria/common/ServiceInvocationContext.java

```diff
@@ -149,6 +154,116 @@ public final EventLoop eventLoop() {
+     */
+    public final EventExecutor contextAwareEventExecutor() {
+        return new ServiceInvocationContextAwareEventExecutor(this, eventLoop());
+    }
+
+    /**
+     * Returns a {@link Executor} that will execute callbacks in the given
+     * {@code executor}, making sure to propagate the current invocation context
+     * into the callback execution. It is generally preferred to use
+     * {@link #contextAwareEventExecutor()} to ensure the callback stays on the
+     * same thread as well.
+     */
+    public final Executor makeContextAware(Executor executor) {
+        return runnable -> executor.execute(makeContextAware(runnable));
+    }
+
+    /**
+     * Returns a {@link Callable} that makes sure the current invocation context
+     * is set and then invokes the input {@code callable}.
+     */
+    public final <T> Callable<T> makeContextAware(Callable<T> callable) {
+        ServiceInvocationContext propagatedContext = this;
+        return () -> {
+            boolean mustResetContext = propagateContextIfNotPresent(propagatedContext);
+            try {
+                return callable.call();
+            } finally {
+                if (mustResetContext) {
+                    removeCurrent();
+                }
+            }
+        };
+    }
+
+    /**
+     * Returns a {@link Runnable} that makes sure the current invocation context
+     * is set and then invokes the input {@code runnable}.
+     */
+    public final Runnable makeContextAware(Runnable runnable) {
+        ServiceInvocationContext propagatedContext = this;
+        return () -> {
+            boolean mustResetContext = propagateContextIfNotPresent(propagatedContext);
+            try {
+                runnable.run();
+            } finally {
+                if (mustResetContext) {
+                    removeCurrent();
+                }
+            }
+        };
+    }
+
+    /**
+     * Returns a {@link GenericFutureListener} that makes sure the current invocation
+     * context is set and then invokes the input {@code listener}.
+     */
+    public final <T extends Future<?>> GenericFutureListener<T> makeContextAware(
+            GenericFutureListener<T> listener) {
+        ServiceInvocationContext propagatedContext = this;
+        return new GenericFutureListener<T>() {
+            @Override
+            public void operationComplete(T future) throws Exception {
+                boolean mustResetContext = propagateContextIfNotPresent(propagatedContext);
+                try {
+                    listener.operationComplete(future);
+                } finally {
+                    if (mustResetContext) {
+                        removeCurrent();
+                    }
+                }
+            }
+
+            // Delegate equality checks to the original listener so removeListener works.
+            @Override
+            public boolean equals(Object obj) {
+                return listener.equals(obj);
```

Comment:
I see - in that case maybe I should get rid of ContextAwareFuture and friends, as they have the problem of not supporting removeListener at all, they would have been convenient but I don't want them to be missing functionality. We can just document the fact that users will need to add listeners using makeContextAware(listener). What do you think?